### ISSUE TITLE
Support Module Source Hierarchy

### ISSUE THREAD

**Issue by desruisseaux** [gradle/gradle#25974]
### Expected Behavior

Since Java 9, the Java compiler and other JDK tools support two different ways to organize the source code: _Package Hierarchy_ and _Module Source Hierarchy_ (see [Directory Hierarchies](https://docs.oracle.com/en/java/javase/20/docs/specs/man/javac.html#directory-hierarchies) in `javac` documentation). However Gradle currently supports only _Package Hierarchy_, which was the only layout supported by all Java versions up to 8. Supporting _Module Source Hierarchy_ in a Gradle project is possible, but difficult. It took us many weeks of effort, and the result is summarized in a page titled [Java Module Source Hierarchy in a Gradle sub-project](https://geomatys.github.io/draft/Modularization.html).

Issues #25954 and #25962 were filled as small steps for making easier for developers to hack a Gradle build, but it would be very appreciated if Gradle had out-of-the-box support for Module Source Hierarchy. It would require some way to tell to all Gradle tasks that a sub-project uses Source Module Hierarchy, then improvements in at least the following tasks:

* `compileJava` and `compileTestJava` for using the `--source-module-path` option.
* `test` for using the `--patch-module` and `--add-exports` options.
* `jar` for generating one JAR per JPMS module instead of one JAR for the whole sub-project.
* `generatePomFile` for filtering the dependencies on a per JPMS module basis.
* `javadoc` in a way similar to `compileJava`.

We wrote Java code for supporting all the above in an experimental build of the Apache SIS project. It currently contains some hard-coded values because I do not have enough Gradle knowledge for doing it properly. But I would be happy to contribute with that code if Gradle team can provide guidance.

### Context

Few developers are aware of the _Module Source Hierarchy_ offered by Java tools since JDK 9, presumably because this layout is unsupported by both Gradle and Maven. However Module Source Hierarchy is used by the OpenJDK project itself and has some great advantages [described in above-cited page](https://geomatys.github.io/draft/Modularization.html#advantages). To summarize, it gives to developers an extra level of flexibility for organizing their modules, enables more compile-time checks of cross-module documentation, facilitates aggregated javadoc and aggregated annotation processing, facilitates the reuse of test fixtures between modules, and more. Module Source Hierarchy can help developers to write better code, it is unfortunate that developers can hardly use it.

**Comment by jvandort** [comment]
As mentioned in the other issue, the [JavaModuleDetector](https://github.com/gradle/gradle/blob/1045229ba88d7165101cba22d99f5ad1465bcd81/subprojects/core/src/main/java/org/gradle/internal/jvm/JavaModuleDetector.java#L38) handles detecting whether a given dependency should be placed on the modulepath or the classpath, based on the presence of `module-info` files or `Automatic-Module-Name` attributes. This would be the place where I think we need to make updates. 

You mentioned you'd be potentially interested in contributing a fix for this issue. I believe the solution to this problem should be entirely contained to the above-linked class. Along with the actual changes to that class, we'd also want to update [JavaModuleDetectorTest](https://github.com/gradle/gradle/blob/7a0de62fa88b31ced22971f6ab7e7e0686218d59/subprojects/core/src/test/groovy/org/gradle/internal/jvm/JavaModuleDetectorTest.groovy#L28). We'd also want to add integration tests: 
- [JavaModuleExecutionIntegrationTest](https://github.com/gradle/gradle/blob/7a0de62fa88b31ced22971f6ab7e7e0686218d59/subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/jpms/execution/JavaModuleExecutionIntegrationTest.groovy#L4): Verify `application` plugin execution
- [JavaModuleCompileIntegrationTest](https://github.com/gradle/gradle/blob/7a0de62fa88b31ced22971f6ab7e7e0686218d59/subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/jpms/execution/JavaModuleExecutionIntegrationTest.groovy#L19): Verify `java-library` plugin compilation
- Any other subclasses of [AbstractJavaModuleIntegrationTest](https://github.com/gradle/gradle/blob/7a0de62fa88b31ced22971f6ab7e7e0686218d59/subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/jpms/AbstractJavaModuleIntegrationTest.groovy#L30) which may also need additional verification (javadoc, testing, at the minimum)

**Comment by desruisseaux** [comment]
Thanks. I will take a look at those classes and come back if I have questions. It may take a few weeks.

**Comment by jvandort** [comment]
After reading this issue again, it seems I misinterpreted the initial issue and this is much larger scoped than I had originally thought. I expect that for some of your listed tasks, implementations may be self-contained enough for external contribution, but for other more complicated tasks, internal design and implementation will likely be required. 

Given that, in the next coming weeks, we will be discussing internally our plans for JPMS and if we will be able to dedicate resources towards improving Gradle's compatibility any time in the foreseeable future. 

In any case, I think this issue is too complex -- containing too many acceptance criteria -- and should be split up into multiple linked issues so that we can more easily prioritize and track work for each task individually. However, this can happen after we've discussed more internally about priority. 

For now, I think it is important to split off the most important work. Of the above-listed tasks, which do you feel are critical for even basic JPMS support in Gradle? Is there anything listed that, if fixed, can unblock significant use cases for users who wish to develop JPMS applications and libraries in Gradle?

**Comment by desruisseaux** [comment]
Indeed it is a task of large scope, and I agree that it probably needs internal discussion in Gradle team for a plan. The list of tasks could be reworded in two main groups:

### Recognize module pattern
The `--module-source-path` option of the Java compiler handles the `*`, `{` and `}` characters in a special way (see [Module-pattern form](https://docs.oracle.com/en/java/javase/20/docs/specs/man/javac.html#module-pattern-form) on Oracle's documentation):

```
javac --module-source-path=/foo/*/bar
```

One possible way to support that in Gradle may be to introduce a new `modulePattern` method that can be used in `setSrcDir` like below :

```
sourceSets {
    main {
        java {
            setSrcDirs(modulePattern("src/*/main/java))
        }
    }
    test {
        java {
            setSrcDirs(modulePattern("src/*/test/java))
        }
    }
}
```

The use of `modulePattern` would have the following effects:

* Set some flag inside Gradle telling that module source hierarchy is used. A consequence is that Gradle shall not emit the `--source-path` option anymore to the Java compiler, and emit the `--source-module-path` option instead with the `modulePattern` value verbatim (including the `*`, `{` and `}` characters),
* For other consumer of source file directories that do not understand the module pattern, Gradle could behave _as if_ the source files were expanded like below:

```
sourceSets {
    main {
        java {
            setSrcDirs(listOf(file("src/foo.bar.a/main/java),
                              file("src/foo.bar.b/main/java),
                              file("src/foo.bar.C/main/java),
                              /* etc. /* )
        }
    }
    test {
        java {
            setSrcDirs(listOf(file("src/foo.bar.a/test/java),
                              file("src/foo.bar.b/test/java),
                              file("src/foo.bar.c/test/java),
                              /* etc. /* )
        }
    }
}
```

Above expansion allows for instance a project to be opened in an IDE.

Caveats:
* Make sure that dependencies are on the module-path.
* If compiling only a few Java classes instead of all of them, it may be necessary to add an `--add-modules` option with a comma-separated list of all modules of the project (not dependencies). It tells the compiler to continue to take other modules in account even if none of their files is  recompiled.
* When compiling tests, we need to generate a `--patch-module` option for each module to compile. Values are of the form _module_=_test source directory_.

Some code inspiration may be found [here](https://github.com/Geomatys/sis/blob/b2f314b580bde2317b01bcd6e64a48e2f4c6d33f/buildSrc/src/org.apache.sis.buildtools/main/org/apache/sis/buildtools/gradle/ModularCompilation.java#L110). This link contains hard-coded stuff specific to Apache SIS project, but the main idea can be seen.

### Separation of JAR files
The JAR task needs to be aware that the Gradle sub-project can generate many JAR files (one per JPMS module) instead of one. This is easy to do since `javac` created a sub-directory for each module. We just need to iterate over those directories.

The Maven `pom.xml` file to generate needs a little bit more work because each JPMS module may use only a subset of the dependencies that we declared for the project. So ideally, each Maven `pom.xml` should have the dependencies filtered for containing only those that are used by the JPMS module. Some sources of inspiration may be found in [this file](https://github.com/Geomatys/sis/blob/b2f314b580bde2317b01bcd6e64a48e2f4c6d33f/buildSrc/src/org.apache.sis.buildtools/main/org/apache/sis/buildtools/maven/Generator.java#L196).

**Comment by desruisseaux** [comment]
Yes I was expecting that some kind of specification draft would need to be written. I can volunteer for that, but would need a mentor for both the form (where/how to write) and for saying what is the "gradle way" to design things.

**Comment by desruisseaux** [comment]
Hello @big-guy. I can volunteer for drafting a proposal, but would it be possible to have someone to guide me? Where to write it, which template?

**Comment by desruisseaux** [comment]
@big-guy, @ov7a : this issue is becoming a high priority for us. Without module source hierarchy, the Java compiler produces hundreds of warnings about unknown modules in `opens` or `exports` statements, or about unknown targets in Javadoc links. With module source hierarchy, we are getting complains from exasperated developers about the terrible lack of support in IDE, which get confused by the workarounds described in this issues. I would be happy to contribute with what I can, but I do not know Gradle internal and would need guidance. If there is a "change request" to fill, is there a template I could use? If there is a prototype to write, can someone point me to the code to edit?

Thanks

**Comment by big-guy** [comment]
Hi @desruisseaux,

I read through the [post](https://geomatys.github.io/draft/Modularization.html) you linked to in the description. 

IIUC, you're proposing that if you have multiple modules, instead of organizing them as separate projects like:
```
module.a/src/main/java/...
module.b/src/main/java/...
```
You would instead organize them as a single project with multiple modules:
```
src/module.a/main/java/...
src/module.b/main/java/...
```
This seems to be at odds with what the JDK describes as a Module Source Hierarchy because that seems to follow the first convention (module name and then src/main/java). The post describes a similar, but different convention. I understand that this was just a suggestion in the javac documentation.

My reading of the [original JEP](https://openjdk.org/jeps/201) that introduced these flags is that this was done to support layouts like the JDK uses. I don't think this translates well into general usage. 

There are a few other problems with this approach:
1. A Gradle project typically produces a single "thing" (a library or application). If you have multiple modules, most plugins are going to be broken or ignore everything outside the "main" module unless they also understand something like a module-source-path or analyze each module as a separate group of sources.
2. A user typically wants to configure a single "thing" in a project. If you can have multiple modules in one project, users will either be unable to specify different dependencies for each module or jump through different hoops.

We aren't going to go in this direction. 

That said, I do think our JPMS support is very basic and does the wrong thing. Instead of trying to build all the modules inside a single project, I think we should focus on how we detect when we need to treat something like a module. If there are specific problems here, we can discuss/fix those. 

In the post, one of the motivations for this layout was to better separate modules.  You can do this in another way.  Instead of including all of the modules in one Gradle build (IOW, listing them all in a single settings.gradle), you can create several included builds.

So instead of endorsed representing a single Gradle project with many modules in it, you can have endorsed be a single Gradle build with multiple subprojects.  incubator and optional could be separate builds as well and they can all be brought back together with a single root build that includes them all.

**Comment by desruisseaux** [comment]
Hello @big-guy. Thanks for the reply. I'm aware that we can create a Gradle build with many sub-projects, but this is not equivalent. As far as I know, there is no way to achieve some of the Module Source Hierarchy benefits with Gradle builds. More on it below. But first, a few clarifications:

JEP-201 indeed makes arbitrary choices with their `src/$MODULE/{share,$OS}/classes/$PACKAGE/*.java` scheme. The `src`, `share`, `$OS` and `classes` parts in this scheme are arbitrary, in the same way as `endorsed`, `incubator`, `src`, `main`, `test`, etc. parts in my post were arbitrary. However, what are not arbitrary are the `$MODULE` and `$PACKAGE` parts. Before Module Source Hierarchy, the mandatory scheme was `<anything>/$PACKAGE/*.java`. With Module Source Hierarchy, the mandatory scheme is `<anything>/$MODULE$/<anything>/$PACKAGE/*.java`. The users can choose whatever they want for the `<anything>` parts. But the `$MODULE` and `$PACKAGE` parts are enforced by the `javac` compiler. So in the following comment:

> IIUC, you're proposing that if you have multiple modules, instead of organizing them as separate projects like:
>
> ```
> module.a/src/main/java/...
> module.b/src/main/java/...
> ```
>
> You would instead organize them as a single project with multiple modules:
>
> ```
> src/module.a/main/java/...
> src/module.b/main/java/...
> ```

The `module.a/src/main/java/...` versus `src/module.a/main/java/...` question is not important, and not the real topic of this issue because Gradle or the users can choose whatever they want. The important topic is to have the `module.a` and `module.b` parts handled by `javac` instead of Gradle, for the reasons exposed below.

With Module Source Hierarchy, building a group of modules is still done by a single call to `javac` or `javadoc` tools. This RFE is not a request for Gradle to iterate over the modules and invoke `javac` for each of them. For plugins that delegate their work to JDK tools working on source code, the impact should be small (mostly an adjustment in parameters) . For plugins working on JAR files, there is more work to do because they need to iterate over the JARs. But it still easy for some tasks such as signing. Furthermore, it may be acceptable to have those plugins throwing some kind of `UnsupportedOperationException`, and warn users that not all plugins support Module Source Hierarchy. Having `javac`, `javadoc`, `jar`, publish and signing support could be enough for starting (I can contribute code for them).

> If you have multiple modules, most plugins are going to be broken or ignore everything outside the "main" module

Conversely, if a plugin needs to process many modules together (for example aggregated Javadoc), the current Gradle and Maven layout makes the task difficult for them. So I'm not sure that any of the two approaches is fundamentally more difficult than the other. They put the difficulty in different places.

## Why Gradle sub-projects are not equivalent
As far as I know, Gradle projects, like Maven projects, are acyclic graphs. If module _C_ depends on module _B_ which itself depends on module _A_, it works fine. But there is no way for module _A_ to reference module _B_ or _C_ except by Java reflection. By contrast, Module Source Hierarchy allows a *limited* form of cyclic graphs:

* In `exports` statements of `module-info` files.
* In `opens` statements of `module-info` files.
* In `@see`, `@link` and `@linkplain` Javadoc tags.

For example, module _B_ can have the following `module-info` file:

```
module B {
    requires A;                          // Module A is a dependency of B.
    exports com.foo.api;                 // Public API exported to everyone.
    exports com.foo.internal to A, C;    // Private API exported only to modules A and C.
}
```

In the above, _A_ is a dependency of _B_, while _C_ is a *dependant* of _B_. The reference to A is no problem with Gradle sub-projects. However the reference to C is a problem. As far as I know, there is no way to make the Java compiler aware of the existence of module C without Module Source Hierarchy, because trying to do that with Gradle projects would create a cyclic graphs. By contrast, with Module Source Hierarchy, `javac` manages those cycles as well as it manages cycles between classes or cycles between methods. The same issue applies to `opens` statements as well. If the `-Xdoclint` option is passed to `javac`, then the same issue applies also to `@see` and similar Javadoc tags.

I presume that Oracle's developers were aware of the problem for build tools, because references to dependants like _C_ in `exports` and `opens` statements causes `javac` to emits only warnings instead of a fatal errors as it usually does for unknown symbols. However in medium-sized projects, we quickly get hundreds of warnings.

### The question
Does Gradle wants to allow developers to resolve those `javac` warnings? If yes, then I'm not aware of any way other than supporting Module Source Hierarchy.

**Comment by desruisseaux** [comment]
I believe that this issue has been closed on the basis of a misunderstanding of Module Source Hierarchy, maybe because I mixed unnecessary things in my attempt to explain (in particular changes of `src/main/java` layout convention). To demonstrate above reply, I created [a very small project showing the issue](https://github.com/Geomatys/MavenModulepathBug/tree/cyclic-graph). The project contains 3 modules named _a_, _b_ and _c_ where _a_ is a dependency of _b_, and _b_ is a dependency of _c_. The source files are organized as classical Gradle projects. For example the `build.gradle` file of module _b_ is simply as below (note that there is no change in source directory layout):

```groovy
apply plugin: 'java'
description = 'Module B'
dependencies {
    implementation(project(":a"))
}
```

The `module-info.java` file of module `b` is as discussed in above comment:

```java
module b {
    requires a;
    exports b.api;                  // Package accessible by everyone.
    exports b.internal to a, c;     // Package accessible only by selected modules.
}
```

Forward references like _c_ above are allowed in `exports`, `opens` and Javadoc tags. Actually, if we ignore Java reflection, useful `exports` statements can only contain forward references. Exporting to module _a_ is useless (except if module _a_ does reflection) because _a_ cannot requires module _b_. Forward references are allowed by necessity in `exports` statements, even at the cost of cyclic graphs, but are not allowed in `requires` statements for example.

## Test with classical Gradle module
Run the following commands in a Unix shell (tested with Gradle 8.3 and Java 21):

```bash
git clone https://github.com/Geomatys/MavenModulepathBug cyclic-graph
cd cyclic-graph
git checkout cyclic-graph/gradle-modules
gradle compileJava
```

The compilation succeed, but with the following warning:

```
> Task :b:compileJava
b/src/main/java/module-info.java:4: warning: [module] module not found: c
    exports b.internal to a, c;     // Package accessible only by selected modules.
                             ^
1 warning
```

## Test with Javadoc check enabled
The following commands test a case identical to the previous test, with only the `-Xdoclint` compiler option added in module _a_:

```bash
gradle clean
git checkout cyclic-graph/gradle-modules-doclint
gradle compileJava
```

Now, the compilation fails with a fatal error:

```
> Task :a:compileJava FAILED
a/src/main/java/a/api/A.java:4: error: reference not found
 * @see c.api.C
^
1 error
```

## Test with Module Source Hierarchy
The following commands change absolutely nothing to Java source code and directory hierarchy. This test demonstrates that this issue is not about the `src/main/java` hierarchy convention. The tag only changes Gradle configuration. It removes all `build.gradle` files in modules _a_, _b_, _c_ and replaces them by a single `build.gradle` file in the root directory. The `-Xdoclint` compiler option is still present. This build configuration gives all modules together to the Java compiler, in one single call to `javac`, and lets the compiler manages inter-dependencies between modules.

```bash
gradle clean
git checkout cyclic-graph/module-source-hierarchy
gradle compileJava
```

All Java compiler warnings and errors are gone!

* No more "module not found: c" warning!
* No more "reference not found: see c.api.C" error!

We can rely on compile-time checks of forward references. This safety is the reason why Module Source Hierarchy is needed. I'm not aware of any way to resolve those compiler warnings and errors with classical Gradle modules. If some way is nevertheless possible, I suspect it may be more convolved than relying on compiler build-in support.

So @big-guy, the question is: can you compile this `cyclic-graph` project without compiler warning using Gradle modules? (without disabling warning of course).

**Comment by desruisseaux** [comment]
Ping? I would like to check if this issue was closed by misunderstanding. The argument was that we can do the same thing with Gradle modules, but this is not exactly true as far as I know. To check, if the answer to any of the following question is "yes", then this issue can really be closed:

* Can you compile [this very simple test case](https://github.com/Geomatys/MavenModulepathBug/tree/cyclic-graph) with Gradle modules without Java compiler warnings?
* If no, is it Gradle's decision to not provide support for developers wanting to build modular projects with those warnings resolved?

**Comment by oleg-nenashev** [comment]
Reopening this case for re-triage, as there was a misunderstanding. Discussed it with @desruisseaux at Community over Code. See the test case above

**Comment by makkarpov** [comment]
To further comment on this issue - in my opinion, it should be somehow supported with Gradle's multi-project builds as well.

If you put everything in a single source set (just as test case suggests) - then all of your modules will have same set of dependencies. This is not significant when building standalone JAR files, but is very significant when you intend to publish these to Maven.

**Comment by big-guy** [comment]
Hi @desruisseaux,

We agree that there's something missing here. We'd like Gradle to support JPMS more fully.

Our goal wouldn't be to support module source hierarchy specifically, but we want to solve how Gradle can model circular dependencies  in a module.  We need to do this investigation.  Part of the problem is that modules and projects do not align well when there are circular module dependencies and building multiple modules (within one project) does not align with how Gradle would publish and consume multiple jars. 

I don't know when we will be able to get to this, certainly not before Gradle 9.0.

**Comment by desruisseaux** [comment]
> If you put everything in a single source set (just as test case suggests) - then all of your modules will have same set of dependencies.

Not necessarily. If deploying to Maven Central, we need to generate `pom.xml` file for each module anyway. The code generating those `pom.xml` files can filter the dependencies for including only the ones declared in `module-info`. This is exactly what our custom Gradle plugin for Apache SIS does.

In Maven terminology, a project using Module Source Hierarchy would typically have no `<dependencies>` section (except for test dependencies), but only a `<dependencyManagement>` section. The `<dependencies>` section is generated for each module as the intersection of `<dependencyManagement>` with `module-info`.

**Comment by tkslaw** [comment]
**Preface:** I don't know if this belongs here or in a new issue, or if this has been brought up elsewhere already. I also don't know if there's a workaround.

---

Generating Javadoc with the standard `javadoc` task is currently leaking internal details when developing with JPMS modules. For instance, if you have a sealed hierarchy across packages:

##### `module-info.java`

```
module mylib {
  exports com.example;
}
```

##### `com/example/Greeter.java`

```
package com.example;

import com.example.internal.ConsoleGreeter;

public sealed interface Greeter permits ConsoleGreeter {

  void sendGreeting(String message);
}
```

##### `com/example/internal/ConsoleGreeter.java`

```
package com.example.internal;

import com.example.Greeter;

public final class ConsoleGreeter implements Greeter {

  @Override
  public void sendGreeting(String message) {
    System.out.println(message);
  }
}
```

Then the generated Javadoc leaks the internal `ConsoleGreeter` class. It's package (correctly) does *not* appear in the module documentation. But if you go to the `Greeter` interface's documentation you can see `ConsoleGreeter` in the `permits` section. Worse, it's a *navigable link* that fully documents the class. I attempted to work around this problem in two ways:

1. Exclude the internal source files. This caused the Javadoc generation to fail with a "cannot find symbol" error for `ConsoleGreeter`.

    ```
    tasks.javadoc {
      exclude("**/internal/**")
    }
    ```

2. Pass `-exclude` option to the tool. This had no effect.

    ```
    tasks.javadoc {
      (options as CoreJavadocOptions).apply {
        addStringOption("exclude", "com.example.internal")
      }
    }
    ```

**This problem does not occur when generating the Javadoc with `--module-source-path` and `--module`.** Unfortunately, while I can manually define those options I cannot prevent the Gradle `javadoc` task from passing a list of source files. Setting the task's `source` to an empty file collection/tree leads to the task being skipped due to "empty source". And it's the listing of source files that seems to be the primary cause of this problem.
    
Note I do not believe this is a bug in the JDK's `javadoc` tool. At least not in this specific scenario. Having sealed hierarchies across packages explicitly requires the source to be compiled as a module. Outside of modules, sealed hierarchies are limited to a single package.

**Comment by tkslaw** [comment]
Regarding the source directory structure of Gradle (sub)projects in the context of `--module-source-path`, it doesn't *have* to change. That option supports values in the `<module>=<path>` syntax (e.g., `mylib=src/main/java`). At least as of Java 12; see https://bugs.openjdk.org/browse/JDK-8208609.

Maven 4 looks like it's supporting both `src/main/java` and `src/<module-name>/main/java`. See https://maven.apache.org/plugins/maven-compiler-plugin-4.x/modules.html.

In general, I think Gradle needs a way to specify the Java module name in the build script that all tasks can make use of. Probably in `SourceSet`.

**Comment by desruisseaux** [comment]
Maven 3 tried to support Java modules transparently, apparently with the goal of making modules practically invisible to users. Gradle seems to have followed the path of Maven 3 without really reconsidering whether it was a good idea. In my opinion, this automation attempt was a mistake which may have been the main impediment to Java modules adoption. The Java architects [expressed frustration about inadequate modules support in Java build tools](https://youtu.be/DrF4dCC0daE?amp;t=825) and wondered (near the end of the video) if Oracle will need to provide a build tool themselves.

Good Java modules support cannot be made automagic. Developers need to be able to specify explicitly whether they want their dependencies to be placed on the class-path or module-path. Ideally, build tools should support module source hierarchy. But even if a build tool doesn't want to support module source hierarchy because of fear of changing the way that the tool works, they need to redesign their tools *as-is* Java module hierarchy was going to be supported, eventually stopping before the point where we can compiler many modules at once. It would help to resolve issues like the one describe in above comment. Finally, the build tools will need some way to make easy to specify `--add-reads`, `--add-opens` and `--add-exports` options for testing purposes. Maven 4 invented a `module-info-patch.maven` file for that, but other build tools can invent their own mechanism if they want. In any cases, some changes in users habits are unavoidable.
